Enhance Upsdebugx with the message's debug-level tag

jimklimov opened the pull request:
For debugging output, we want to prepend the debug level so the user can e.g. more easily lower the level (less -D's on command line) to retain just the amount of logging info he needs to see at the moment. Using '-DDDDD' all the time is too brutal and needed high-level overview can be lost. This [D#] prefix can help limit this debug stream quicker, than experimentally picking ;)

Example screenshot (testing code in screenshot is not part of this commit, a copy for reference is just below in the comment):
````
$ ./snmp-ups -DDDDDDDD -a x
Network UPS Tools - Generic SNMP UPS driver 0.99 (2.7.4-109-g0059555)
   0.000000	[D2] Processing CLI option  1 'D' : '(null)'
   0.000018	[D2] Processing CLI option  1 'D' : '(null)'
   0.000025	[D2] Processing CLI option  1 'D' : '(null)'
   0.000033	[D2] Processing CLI option  1 'D' : '(null)'
   0.000103	[D2] Processing CLI option  1 'D' : '(null)'
   0.000110	[D2] Processing CLI option  2 'D' : '(null)'
   0.000117	[D2] Processing CLI option  4 'a' : 'x'
   0.000133	Can't open /usr/local/ups/etc/ups.conf: Can't open /usr/local/ups/etc/ups.conf: No such file or directory
````

The quick-test code was to debug-print something somewhere before we fail because NUT is not installed and set up on the workstation ;)
````
diff --git a/drivers/main.c b/drivers/main.c
index f63d54a..d85cf35 100644
--- a/drivers/main.c
+++ b/drivers/main.c
@@ -509,6 +509,7 @@ int main(int argc, char **argv)
 	upsdrv_makevartable();
 
 	while ((i = getopt(argc, argv, "+a:kDhx:Lqr:u:Vi:")) != -1) {
+		upsdebugx(2,"Processing CLI option %2d '%c' : '%s'",optind, i, optarg);
 		switch (i) {
 			case 'a':
 				upsname = optarg;
````

clepple commented:
@jimklimov I like this in theory, but the macros worry me a bit. I would almost prefer to have [D0] left in for consistency.

Also, in cd10c52 on OpenBSD:
```
../../common/common.c: In function 'upsdebug_with_errno':
../../common/common.c:436: error: #pragma GCC diagnostic not allowed inside functions
../../common/common.c:440: error: #pragma GCC diagnostic not allowed inside functions
```
http://buildbot.networkupstools.org/public/nut/builders/OpenBSD-x64/builds/485/steps/compile/logs/stdio

```
$ cc --version
cc (GCC) 4.2.1 20070719 
```

jimklimov commented:
Would it help if the macros were moved to be around these two functions
wholly?

--
Typos courtesy of GMail on my Samsung Android

19.09.2016 14:54 пользователь "Charles Lepple" <notifications@github.com>
написал:

> @jimklimov <https://github.com/jimklimov> I like this in theory, but the
> macros worry me a bit. I would almost prefer to have [D0] left in for
> consistency.
>
> Also, in cd10c52
> <https://github.com/networkupstools/nut/commit/cd10c529a7cdd84c24fa6c2ad1571c15b8ef28ed>
> on OpenBSD:
>
> ../../common/common.c: In function 'upsdebug_with_errno':
> ../../common/common.c:436: error: #pragma GCC diagnostic not allowed inside functions
> ../../common/common.c:440: error: #pragma GCC diagnostic not allowed inside functions
>
> http://buildbot.networkupstools.org/public/nut/builders/OpenBSD-x64/
> builds/485/steps/compile/logs/stdio
>
> $ cc --version
> cc (GCC) 4.2.1 20070719
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/networkupstools/nut/pull/323#issuecomment-247984951>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABj5lIvg2j7kONzTLeIpjAZX42gMN9hbks5qroX_gaJpZM4J_sk2>
> .
>

jimklimov commented:
The macro approach was seen by me in various projects and posts, so I guess this is a relatively widely-working solution (pragmas for that are not standard though, so there is complexity like this macro monster to implement the feature for certain compilers and may rely on their versions; hopefully for others the macro is a no-op).

I've updated the patch, and verified that the compilation does not crash with syntax errors if I changed the macro definition to not kick in for `__GNUC__` : replaced it in macro tests with `__GNUCx__` locally -- the `-Wvarargs` warnings appear (and a `-Werror` build would die on this - hence all the trickery to essentially request `-Wall -Wvarargs=no -Werror` in GCC speak, but portably).

jimklimov commented:
Thanks for the buildbotnet run :)

The one failing job http://buildbot.networkupstools.org/public/nut/builders/FreeBSD-x86/builds/545/steps/compile/logs/stdio is apparently due to the inability to generate docs (solved in another PR, not part of this branch), so changes about `upsdebugx` and macros are acceptable for that platform.

jimklimov commented:
On a side note, and not focus of this PR, when I enabled `-Werror` for a distcheck, other issues were reported in code I did not touch ;) That's just FYI, I did not look deeper into this.

````
/bin/sh ../libtool  --tag=CC   --mode=compile /usr/lib/ccache/gcc -DHAVE_CONFIG_H -I. -I../../../common -I../include    -I../include -I../../../include -g -O2 -Wall -Wsign-compare -D_REENTRANT -Werror -Wall -Wall -Wsign-compare -D_REENTRANT -MT snprintf.lo -MD -MP -MF $depbase.Tpo -c -o snprintf.lo ../../../common/snprintf.c &&\
mv -f $depbase.Tpo $depbase.Plo
libtool: compile:  /usr/lib/ccache/gcc -DHAVE_CONFIG_H -I. -I../../../common -I../include -I../include -I../../../include -g -O2 -Wall -Wsign-compare -D_REENTRANT -Werror -Wall -Wall -Wsign-compare -D_REENTRANT -MT snprintf.lo -MD -MP -MF .deps/snprintf.Tpo -c ../../../common/snprintf.c  -fPIC -DPIC -o .libs/snprintf.o
../../../common/snprintf.c:556:20: error: conflicting types for built-in function 'pow10' [-Werror]
 static LDOUBLE pow10 (int exp)
                    ^
../../../common/snprintf.c:569:13: error: conflicting types for built-in function 'round' [-Werror]
 static long round (LDOUBLE value)
             ^
../../../common/snprintf.c: In function 'fmtfp':
../../../common/snprintf.c:607:8: error: unused variable 'fracpart' [-Werror=unused-variable]
   long fracpart;
        ^
../../../common/snprintf.c:606:8: error: unused variable 'intpart' [-Werror=unused-variable]
   long intpart;
        ^
   int caps = 0;
       ^
../../../common/snprintf.c: At top level:
../../../common/snprintf.c:556:20: error: 'pow10' defined but not used [-Werror=unused-function]
 static LDOUBLE pow10 (int exp)
                    ^
../../../common/snprintf.c:569:13: error: 'round' defined but not used [-Werror=unused-function]
 static long round (LDOUBLE value)
             ^
../../../common/snprintf.c: In function 'dopr':
../../../common/snprintf.c:278:34: error: 'short int' is promoted to 'int' when passed through '...' [-Werror]
    value = va_arg (args, short int);
                                  ^
../../../common/snprintf.c:278:34: note: (so you should pass 'int' not 'short int' to 'va_arg')
../../../common/snprintf.c:278:34: note: if this code is reached, the program will abort
../../../common/snprintf.c:290:34: error: 'short unsigned int' is promoted to 'int' when passed through '...' [-Werror]
    value = va_arg (args, unsigned short int);
                                  ^
../../../common/snprintf.c:290:34: note: if this code is reached, the program will abort
../../../common/snprintf.c:302:34: error: 'short unsigned int' is promoted to 'int' when passed through '...' [-Werror]
    value = va_arg (args, unsigned short int);
                                  ^
../../../common/snprintf.c:302:34: note: if this code is reached, the program will abort
../../../common/snprintf.c:316:34: error: 'short unsigned int' is promoted to 'int' when passed through '...' [-Werror]
    value = va_arg (args, unsigned short int);
                                  ^
../../../common/snprintf.c:316:34: note: if this code is reached, the program will abort
cc1: all warnings being treated as errors
make[3]: *** [Makefile:593: snprintf.lo] Error 1
````

jimklimov commented:
Reading a bit more about varargs, possibly the initial PR code had a logical error (some test builds annoyed me with unreadable debug messages). One issue was the use of `fmt2[]` defined inside a newly added clause of `upsdebugx()` (so it was freed at the end of clause and in fact sometimes recycled when we went to `vupslog()`). Also it seems that the original `fmt` variable's address better stay in place for varargs inspection, but its value (pointed-to buffer) may be reassigned - so no more `-Wvarargs=error` problems and no more magic pragma/macros are needed.

aquette commented:
@jimklimov : merged with a minor adjustment: c++ style comments were replaced by c style, as per http://networkupstools.org/docs/developer-guide.chunked/ar01s03.html#_portability
Please keep it in mind for the upcoming PRs ;)

clepple commented:
> On a side note, and not focus of this PR, when I enabled -Werror for a distcheck, other issues were reported in code I did not touch ;) That's just FYI, I did not look deeper into this.
> 
@jimklimov out of curiosity, how did you get the common/snprintf.c code to be built? Usually this is only included if the OS does not include snprintf already (per ./configure tests), and snprintf() has been around for a while.

If there is a bug in the autoconf code, please open a new issue with details on that, so we don't forget.
